client/protocol.go

```diff
@@ -0,0 +1,413 @@
+				if err != nil {
+					logrus.Debug(err)
+				} else {
+					isReceivedSuccessRes = 2
+					okCnt++
+				}
+			default:
+				logrus.Warn("unknown command result type")
+			}
+		case *curpapi.ProposeResponse_Error:
+			logrus.Debug(proposeResp.Error)
+		default:
+			okCnt++
+		}
+
+		if isReceivedLeaderReq && okCnt >= superQuorum(len(c.connects)) {
+			logrus.Info("Fast round success. ProposeID:", cmd.ProposeId)
+			if isReceivedSuccessRes == 1 {
+				return &ProposeResponse{
+					CommandResp:        &cmdRes,
+					IsFastRoundSuccess: true,
+				}, nil
+			}
+			if isReceivedSuccessRes == 2 {
+				return &ProposeResponse{
+						IsFastRoundSuccess: true,
+					}, ProposeError{
+						ExecuteError: &proposeErr,
+					}
+			}
+		}
+	}
+
+	logrus.Info("Fast round fail. ProposeID:", cmd.ProposeId)
+	return &ProposeResponse{
+		IsFastRoundSuccess: false,
+	}, nil
+}
+
+// The slow round of Curp protocol
+func (c protocolClient) slowRound(cmd *xlineapi.Command) (*ProposeResponse, error) {
+	logrus.Info("Slow round start. Propose ID: ", cmd.ProposeId)
+
+	leaderId := c.fetchLeader()
+	var syncResp xlineapi.SyncResponse
+	var commandResp xlineapi.CommandResponse
+
+	conn, err := grpc.Dial(c.connects[*leaderId], grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithIdleTimeout(c.timeout.idleTimeout))
+	if err != nil {
+		logrus.Warn(err)
+		return nil, err
+	}
+	protocolClient := curpapi.NewProtocolClient(conn)
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.timeout.waitSyncedTimeout)
+	defer cancel()
+	res, err := protocolClient.WaitSynced(ctx, &curpapi.WaitSyncedRequest{ProposeId: cmd.ProposeId})
+	if err != nil {
+		logrus.Warn(err)
+		return nil, err
+	}
+	waitSyncedResponse := res
+
+	switch syncResult := waitSyncedResponse.SyncResult.(type) {
+	case *curpapi.WaitSyncedResponse_Success_:
+		err := proto.Unmarshal(syncResult.Success.AfterSyncResult, &syncResp)
+		if err != nil {
+			return nil, errors.New("unmarshal syncResp fail")
+		}
+		err = proto.Unmarshal(syncResult.Success.ExeResult, &commandResp)
+		if err != nil {
+			return nil, errors.New("unmarshal commandResp fail")
+		}
+		logrus.Info("Slow round success. Propose ID:", cmd.ProposeId)
+		return &ProposeResponse{
+			SyncResp:    &syncResp,
+			CommandResp: &commandResp,
+		}, nil
+	case *curpapi.WaitSyncedResponse_Error:
+		logrus.Info("Slow round success. Propose ID:", cmd.ProposeId)
+		return nil, CommandSyncError{
+			CommandSyncError: syncResult.Error,
+		}
+	default:
+		logrus.Info("Slow round error. Propose ID:", cmd.ProposeId)
+		return nil, errors.New("unknown synced response")
+	}
+}
+
+// Send fetch leader requests to all servers until there is a leader
+// Note: The fetched leader may still be outdated
+func (c protocolClient) fetchLeader() *uint64 {
+	fetchClusterRespCh := make(chan *curpapi.FetchLeaderResponse)
+	var clients []curpapi.ProtocolClient
+	var leader *uint64
+	var maxTerm uint64 = 0
+	okCnt := 0
+	majorityCnt := len(c.connects)
```

Comment:
The majorityCnt should be `len(c.connects)/2+1)`

Reply:
done